option i18n.notrans for language alias of C locale

wolfmanx opened the pull request:
Correctly terminate the translator fallback chain for the _C_ locale (as _gettext.find() does). Use config option i18n.notrans as a language code alias for the _C_ locale. E.g., when there is no _en_ translation file, since all strings are already in English,

```ini
i18n.notrans = en
```

terminates the translators for an Accept-Language header of

```
Accept-Language: ru,en,fr,de
```

after _ru_:

```python
supported_languages = ['ru']
```

instead of setting up the translator incorrectly as:

```python
supported_languages = ['ru', 'fr', 'de']
```

wolfmanx commented:
This is really just a terse but precise bug report. I'm happy for the pull request to be rejected, as long as the underlying faulty behavior is fixed. No attribution to me is necessary.

amol- commented:
If I get it right, this just avoids having to run `python setup.py init_catalog -l en` if i18n is enabled but `en` is already the language in which the template is originally written into. 
Is my understanding correct?

wolfmanx commented:
You are correct that the problem also appears in the case, where a language (i.e. _en_) is equivalent to the untranslated _C_ locale. 

Generating the _en_ translation file was my first attempt to solve the problem ;-) However,  `python setup.py init_catalog -l en` generates a translation file, where all translations are **empty**. I.e., none of the ID strings are defined which  means all of the ID strings fall through to the next language. the effect is the same, as if there was no translation file to begin with.

The next step was to write a script to replace all empty translations with the respective ID string, which actually works. However, the kallithea maintainers (rightfully) do not want to distribute a superfluous language file. The script is also a real ugly adhoc hack.

So I finally invested the time to look for the actual root of the problem. Besides the fall-through problem for e.g. _en_, there is the underlying problem for the _C_ locale:

`gettext.find` returns a chain of language files. Since _C_ does not have a language file, the language chain is truncated. I.e. 

```python
>>> gettext.find(languages='fr C de ru'.split)
['fr']
```

Single-stepping over the languages with `gettext.find`, assuming that an empty response means "there is no translation file" instead of "either there is no translation file or it is the _C_ locale, which terminates a locale chain, whether there is a _C_ translation or not", is simply wrong:

```python
>>> _get_translator('fr C de ru'.split)
['fr', 'de', 'ru']
```

Checking whether `gettext.find` and `_get_translator` return the same set of languages would probably be a good test case.

The modifications in this patch solve the problem for the _C_ locale. I think, trying to solve this by actually installing a _C_ translation file would be quite hilarious, if it were even possible (there is actually no attempt made by `gettext.find` to locate a _C_ translation file; the list is just truncated).

So once this bug was fixed, defining a configuration value `180.notrans` for a language that is equivalent to the _C_ locale is just a logical extension.

I didn't think it made too much sense to produce a polished patch, since there are open questions like:

1. Should the value `i18n.notrans` have a different name like `i18n.c_equiv`?
2. Should `i18n.notrans` be a list of languages?
3. Should the equivalent languages be replaced by `C` in a separate preprocessing pass?

However, should the patch quality be the  only blocking point, I am willing to beautify it, since I already have more time invested than planned.

wolfmanx commented:
Sorry, wrong button.

amol- commented:
Ok, I was trying to reason about this change, and I don't think I yet have a clear idea.

About `"C" POSIX` Support, I'm convinced that's wrong in the context of a web framework.
While it makes sense for `gettext` in the context of a local desktop application, here we are talking about explicitly resolving an `Accept-Language` header.

`Accept-Language` is explicitly defined as supporting ISO-639-1 ( https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes ) that nowhere recognises `C` as a valid locale.

Also such header is explicitly described as `can be used by user agents to indicate the set of natural languages that are preferred in the response` with `Accept-Language: da, en-gb;q=0.8, en;q=0.7 would mean: "I prefer Danish, but will accept British English and  other types of English".`

Which does describe the exact behaviour that TurboGears is providing, and should never stop if a translation is not found. So "stopping language resolution when C is found" would be simply wrong in the execution of the `Accept-Language` resolution from any point of view.

On the other side, I do get that there is some convenience in not having to fill translations for a language that is the one your template is already written into, and thus that language should be treated as having all strings translated by themselves.

Truncating the language resolution when more languages were specified in `Accept-Language` can also be confusing. So I feel that a more obvious solution might be in having a `SelfTranslations` class that subclasses `NullTranslations` by disabling the `add_fallback` method so that when the list of languages is printed it's explicitly listed that for `en` you used ` SelfTranslations` and that's why you don't get any further language.

At that point I would be ok with having a `i18n.self_translated_langs` option with a list of languages for which instead of going through the standard `.mo` loading flow you just create a `SelfTranslations` object.

wolfmanx commented:
If I understand this correctly, the languages in `i18n.self_translated_langs` would essentially behave as if the translator chain had been cut. So, a value of

``` ini
i18n.self_translated_langs = c, en
```

would behave exactly like the above patch. In that case, it is a viable solution.

The rest is just some opinions ...

However, it is inefficient to build an entire translator chain instead of cutting it. Since this process is performed for all requests with an Accept-Language header, it is still worth considering the cutting solution.

`gettext.find()` does not operate in the scope of the HTTP standard but rather in the scope of POSIX locales. The recognition and shortcircuit evaluation of the 'C' locale is hardcoded into the Python module and must be considered, when calling `gettext.find` (which currently does not happen, but could be realized by the application with `i18n.self_translated_langs`).

IMHO, the HTTP protocol essentially falls short with the narrow definition of Accept-Language, completely ignoring all locale issues (see e.g. [Accept-Language used for locale setting](https://www.w3.org/International/questions/qa-accept-lang-locales.en)). Therefore, I do not think that the "standards compliance" argument should be made on the framework level. It is rather the final decision of the web application in which way to handle the header. (A perfectly good use case for the C locale can always be found in the test arena).

kiilerix commented:
Let me try to summarize and frame this whole problem (from the perspective of Kallithea):

gettext style translations have strings in a source language (usually English), and translation to a language is file with mapping from source language strings to strings translated to the language. If no translation is available, the source lange string is used directly.

Translations are generally found at runtime by looking for .mo files in subfolders of a given location, with target language as folder name. Users can requst a language. If it exists, strings are translated, with default of using source langage. If the language doesn't exist, the default of source language is used too.

Nowhere is it declared what the source language is. That seems like a flaw in the gettext scheme. But usually it doesn't matter much: It will either be the language the user asked for, or the user asked for something that doesn't exist ... and will get the language anyway, no matter what it is. Also, arguably, the source language is already "hardcoded" in the app, and the name of the source language is thus something the app can provide when initializating the translation code (which conceptually also will be where it is specify where translations are to be found).

Http requests specify a priorited list of languages in the Accept-Language header. It could make sense to just use the first translation from this list where the application happens to have a translation file - no matter how complete it is. Another interpretation (which is the current TG implementation) is that each string is tried with all available prioritized languages before falling back to default.

That is where we see the problem with Kallithea: If the source language (which by definition has 100% coverage) is in the prioritized language list, nobody will notice (because the source language is unknown), and the search for translations will continue to languages *after* the one with 100% coverage.

This problem (and solution) is also described in https://github.com/cdent/paste/blob/3.2.3/paste/wsgiwrappers.py#L89 . It is claimed in https://paste.readthedocs.io/en/latest/future.html#to-deprecate that "All the functionality is already in WebOb" ... but this particular thing doesn't seem covered by WebOb. TG uses WebOb. So perhaps the problem really should be solved in WebOb?

Anyway: https://github.com/TurboGears/tg2/blob/tg2.4.2/tg/request_local.py#L36 seems to be doing pretty much the same thing, only with less clear documentation / explanation.
So apparently, the i18n.lang *must* be set to translation source language? It is not really something that makes sense to have customizable in the .ini , but something the app should hardcode (or at least provide as default if not set in the .ini).

But also, to avoid "No translation file found", TG requires an (empty) translation for the source language? That seems like a TG flaw (which apps using TG must work around until it has been fixed in all TG versions they support)?

In Kallithea, https://kallithea-scm.org/repos/kallithea/changeset/f2f7a8c1281e and https://kallithea-scm.org/repos/kallithea/changeset/8931078f70db were thus not spot-on.
We can solve/workaround the problem with https://kallithea-scm.org/repos/kallithea-incoming/changeset/ee69feb67613dceb1e5efd670570d34a3ceabf86 . Does that make sense?

Also:

Standard gettext.py kind of handles this problem in https://github.com/python/cpython/blob/v3.8.1/Lib/gettext.py#L562 , assuming that the source language is called "C". It could be used, by renaming any request for the source language to be a request for C. The problem is made more complex by how TG re-implement some parts of what already is in stdlib gettext. That seems unfortunate, and it is not entirely clear why. If stdlib has severe shortcomings, it should be made clear in separate code that can be reused between projects (like WebOb?) and/or eventually upstreamed.

But also, since most applications and sites use English as source language, it is a bit odd and rare to have anything after English in Accept-Launguage. While the problem is real, it is also rare and easy to work around.

wolfmanx commented:
As for _rare_ and _easy_ ...

> [...] sites use English as source language, it is a bit odd and rare to have anything after English in Accept-Launguage.

As programmer I prefer English, as native German speaker I prefer German. 
Most of the time I am a programmer and I could not care less for the German version of a web site, so for me having English before German is neither odd nor rare. But sometimes I have to assist my German wife or a German customer, so I keep shuffling the language priorities. Indian colleagues do the same thing, since I do not speak any Indian language, we have to meet on common ground.

When I set up my first Kallithea instance, I did not conciously notice that the program displayed German pages. But when I explained Kallithea to an Indian colleague who uses a German PC, he asked after 5 minutes whether the pages could be displayed in English. When setting the language priorities, I realized, that I already had the appropriate setting without the desired effect.

> While the problem is real, it is also rare and easy to work around.

In the specific case of Kallithea, I have to either

- generate a fake English translation file or
- patch the TG source code or
- monkey-patch the TG module.

It may be a matter of opinion whether these options are easy, but they are definitely not _easy_ to live with. Since each programmer in my company gets his own instance of Kallithea, the deployment becomes unnecessarily complex (I would really like to stop using rsync(1) for internal distribution).

wolfmanx commented:
I finally got around to testing:

> We can solve/workaround the problem with <https://kallithea-scm.org/repos/kallithea-incoming/changeset/ee69feb67613dceb1e5efd670570d34a3ceabf86> . Does that make sense?

This workaround behaves properly for the source language problem.  However, I'd rather not hack up the empty language translation file with a printf(1) command. The result of

``` sh
python2 setup.py init_catalog -l en
python2 setup.py compile_catalog -l en
```

works fine. Distributing an empty `i18n/en/LC_MESAGES/kallithea.po` file with the default setting of

``` ini
i18n.lang = en
```

seems to be a good failsafe solution.

> The problem is made more complex by how TG re-implement some parts of what already is in stdlib gettext. That seems unfortunate, and it is not entirely clear why.

The `supported_languages` list cannot be generated otherwise, because `gettext.find` returns a list of language files, not a list of languages.

If no action is taken in TG, it should at least be documented clearly how to handle such a source language problem. Specifically the cutting of the language list from the Accept-Language header based on the setting of `i18n.lang` should be pointed out (<https://github.com/TurboGears/tg2/blob/tg2.4.2/tg/request_local.py#L36>).

kiilerix commented:
Thanks for confirming the Kallithea approach works and seems reasonable given your understanding. It would be nice to confirmation from TG upstream too.

It might be more "standard" to use init_catalog / compile_catalog, but it is also bigger and include things we might not want - such as year.
But we just want a sentinel value. So one goal is to make it as minimal as possible. It could also be created based on gettext.py using:
  python2 -c "import struct,gettext;file('kallithea/i18n/en/LC_MESSAGES/kallithea.mo', 'wb').write(struct.pack('<5I', gettext.GNUTranslations.LE_MAGIC, 0, 0, 0, 0))"
But the best solution would be to change TG so we don't need the dummy file at all.

Also: While i18n.lang "works" if used "correctly", it seems that it is a bit overloaded to be used for two kinds of "fallaback":

1. The language that always should be tried *before* falling back to no translation (a language that always is appended to the list of languages)
2. The language you get from the fallback of no translation (which is where it should stop searching in the list of languages).

It would perhaps be simpler and cleaner if TG extracted the list of langauges from the list of translation files - either from looking at paths, or from looking inside the files.

amol- commented:
Compiling an empty catalog and setting it as the default language, has been the de-facto practice in TG for the approach of supporting the native template language when i18n is enabled. But I do recognise that a more convenient way that doesn't require making an empty catalog is necessary.

That's why I'm ok having an option that injects identity translators (translators that return themselves) instead of loading them from po files. 

What I think it's out of scope it's just caring for `C` locale. Localization is a far more complex problem than translation, and tg.i18n is supposed to provide support for translation only. And it doesn't exist anything like a "C" language. Overall localization problem in TG is usually solved in conjunction of Babel, not by TG alone.

If you don't have time to adapt the change to just return a `TGI18NIdentityTranslator` in `_get_translator` for a set of specified languages I can easily work on that myself as soon as I have some spare time. But I can't guarantee when that will happen.

kiilerix commented:
I must say I don't understand the need for a TGI18NIdentityTranslator when TG already has support for truncating the language list, and it seems like it just could fall through to the end of the list instead of adding the "empty" language (and thus needing an empty .mo file.

The only reason I mentioned `C` is because gettext.py hardcode 'C' as fallback. That might be "wrong", but it seems like it will work just fine if apps before invoking gettext.find rename references to the source language to reference C instead. That might not be relevant if not using gettext.find.

amol- commented:
Because truncating the language list is an hack that allows you to achieve the goal (having a language that translates to itself) in an unclear and unexplicit way instead of explicitly stating what's going on for anyone debugging a webapp / staring at that code.
And TG has already a long enough history of unclear hacks and side-effects :D

wolfmanx commented:
> If you don't have time to adapt the change to just return a TGI18NIdentityTranslator in _get_translator for a set of specified languages [...]

OK, I will give it a try. It is definitely better than the `i18n.lang` workaround.

wolfmanx commented:
This proposal works as expected without a translation file. Languages matching the space or comma separated list of languages in `i18n.self_translated_langs` are instantiated as objects of class `TGI18NIdentityTranslator` and added to the list of translation files. The function `_translator_from_mofiles` is modified to also accept instances of `NullTranslations` as is.

amol- commented:
As I had started it, I pushed the branch that I made with support to `i18n.native` option (I thought it was a better name in the end than `i18n.self_translated_langs`).
You can see it at https://github.com/TurboGears/tg2/compare/development-i18n

It's fairly similar in the end to the code you updated your branch too, but adds tests and uses TG converters to handle option instead of parsing it in `get_translator`.

I'm willing to merge it with you as the contributor, so let me know if you see anything that doesn't work or is wrong.

wolfmanx commented:
The logic looks good. The test are correct. Don't worry about attribution to me., just go ahead and merge the changes. Thanks!

kiilerix commented:
For the record:

This issue got resolved with TurboGears 2.4.3 with bf240a8fb4fc . But that also caused regressions for some use cases set up to work around the previous behaviour: It is now necessary to set i18n.native .

It also seems a bit unfortunate that even though for example `en_US` would use the `en` translation file if there is no `en_US` translation, it will only be recognized as native if the full `en_US` is in `i18n.native`.

amol- commented:
> It also seems a bit unfortunate that even though for example `en_US` would use the `en` translation file if there is no `en_US` translation, it will only be recognized as native if the full `en_US` is in `i18n.native`.

That seems a reasonable improvement, if you are interested in contributing a PR I would be glad to review it. 2.5 will be mostly removing deprecated behaviours, so it might be a good time to introduce that kind of change.
